cpp/src/arrow/io/interfaces.cc

```diff
@@ -34,8 +34,9 @@ Status InputStream::Advance(int64_t nbytes) {
   return Read(nbytes, &temp);
 }
 
-util::string_view InputStream::Peek(int64_t ARROW_ARG_UNUSED(nbytes)) const {
-  return util::string_view(nullptr, 0);
+Status InputStream::Peek(int64_t ARROW_ARG_UNUSED(nbytes), util::string_view* out) {
+  *out = util::string_view(nullptr, 0);
+  return Status::OK();
```

Comment:
Why not return `NotImplemented` here now that we're returning a Status?

Reply:
Well, it didn't fail before. Do you think it should fail?

Reply:
Because it's not able to peek (it returns an empty string_view with a nullptr). Previously we weren't returning a Status, so we couldn't fail explicitly.

Reply:
done